fix strategy hashing arg mismatch

zpcore opened the pull request:
Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159289



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

zpcore committed:
fix strategy hashing arg mismatch

[ghstack-poisoned]

zpcore committed:
Update on "fix strategy hashing arg mismatch"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore committed:
Update on "fix strategy hashing arg mismatch"


Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=-1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

wconstab commented on test/distributed/tensor/test_op_strategy.py:
```diff
@@ -607,5 +607,29 @@ def test_tuple_replicate_strategy_placement(self):
+        with op_strategy_context(torch.ops.aten.sort.default, replicate_op_strategy):
+            # intentionally do not supply `schema_info=RuntimeSchemaInfo(1)`
+            torch.sort(sharded_dtensor, dim=0)  # sort each column
+            out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
```
confusion on what you're trying to test:

if you force the sort op to use the replicate strategy, then shouldn't the strategy return the same sharding no matter what input you give?  (How would this test ever fail?)

zpcore replied:
Yes, the intention is to let sort return the same sharding strategy regardless of the input in this test. When the same strategy triggered, original code will output the same result if call sequentially:
```
torch.sort(sharded_dtensor, dim=0)  # <----- insert the cache key
torch.sort(sharded_dtensor, dim=1)  # <----- read from cache and reuse dim=0, which is incorrect.
```

wconstab commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -307,8 +306,11 @@ def redistribute_local_args(
+                # [args/kwargs]_schema that is not DTensorSpec from
+                # op_info.schema instead of
+                # info.output_sharding.redistribute_schema.
+                new_local_args.append(arg_spec)
```
iiuc, your claim is that `arg_spec` comes from the current runtime call, while `reshard_arg_spec` comes from the cache from the first time someone called DTensor with the same cache-key?

let me think about this for 2 cases

1) `arg_spec` was not part of the cache key for this op: then using the value copied from the cache seems wrong and this seems to fix it
2) `arg_spec` was part of the cache key: then it should have worked with the old code, but the new code is fine too

does that sound right?  well, if so, i'm surprised we didn't get bit by this bug before!

zpcore commented:
Those view ops failed in the test because in https://github.com/pytorch/pytorch/blob/59e261bbd86b173b6cf318d392a613ca3efdbe10/torch/distributed/tensor/_sharding_prop.py#L370-L380, the modified args are stored in suggestion_schema. So in those cases, we need to use `suggestion_schema`.

wconstab commented:
so, the explanation and the fix seems good to me.  But the test failures look suspicious to me. 

e.g. layer_norm_backward failing with  `RuntimeError: shape '[20, 5, 10]' is invalid for input of size 250` - have you triaged this one yet?  I can review again after you get the tests all passing.

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore commented:
> so, the explanation and the fix seems good to me. But the test failures look suspicious to me.
> 
> e.g. layer_norm_backward failing with `RuntimeError: shape '[20, 5, 10]' is invalid for input of size 250` - have you triaged this one yet? I can review again after you get the tests all passing.

Those invalid for input error are caused by those view related ops. Here is the inconsistency:
1) view related ops updated the input args using the output_sharding.redistribute_schema
2) if we use output_sharding.redistribute_schema, the nontensor args will be from the caching, which is not what we want.

To be consistent, I updated the nontensor args in redistribute_schema to follow op_info.flat_args_schema. Also I moved view ops related update outside of the `propagate_op_sharding_non_cached`, where I don't think we should cache those updated stride values.

wanchaol reviewed (commented):
I think it's worth discussing how to solve this, my concern is still that this change incurs some CPU overhead and I think we could think about how to avoid it.

I feel a better way would be, just enforce each op registered to mark every scalar argument for an op to have static_argnum, where they would invalidate cache and run things correctly. This is somewhat already done in `register_op_sharding` path, but for our natively implemented op we manual tag it as it's more accurate.

zpcore commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -149,6 +149,11 @@ def register_op_strategy(
+            if schema_info is not None:
+                self.op_to_schema_info[op_overload] = schema_info
+            elif op_overload in self.op_to_schema_info:
+                del self.op_to_schema_info[op_overload]
```
This is useful if a we want test an op with different op strategies. Some strategy may have op_to_schema_info, some may not.  An example would be:
```
register_op_strategy(op_x,  strategy1, schema_info)
...
register_op_strategy(op_x,  strategy2) <----  need to remove the schema_info from strategy1 we registered previously
...
```

wanchaol replied:
Hmmm this is the problem, why we want to test an op with a different op strategy, but without passing in schema_info? IMO for the same op, the semantic of schema info should be either explicitly pass by the time you register op strategy (error if not passing in), or we should assume they are the same? I couldn't think of a case where the schema info be different for the same op but with different op strategy, can you explain?

zpcore replied:
> I couldn't think of a case where the schema info be different for the same op but with different op strategy

This can be a case:
```
register_op_strategy(aten.sort.default, dont_shard_on_sort_dim, schema_info=RuntimeSchemaInfo(1))
...
register_op_strategy(aten.sort.default,  replication_strategy, schema_info=None)
...
```

There is no way we can update the schema_info if we don't delete the original schema_info. However, I managed to remove the change since `op_strategy_context` helps us clear `propagator.op_to_schema_info` dict(as a hack I suppose). So this is not the issue for this test anymore.

zpcore replied:
Yes, now I just changed the util. I think this is only useful for testing.

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -298,19 +303,61 @@ def spec_to_strategy(spec: object) -> object:
         op_info.output_sharding = output_sharding
+        # handle view op and make redistribute_schema in output_sharding the
+        # source of truth
+        self._maybe_update_redistribute_schema(op_info)
```
I figured out a way to remove this function. I have to add an extra bool var `use_val_from_redistribute_schema` to tell the dispatcher whether to use args val from redistribute_schema.args_schema or orignal op_info.schema. This should get rid of the overhead.

zpcore replied:
By the way, do you know if there exists any benchmark regarding the dispatch overhead measurement? I think it can be helpful to add some tests in the CI to prevent introducing extra overhead.

wanchaol replied:
You can look at this https://github.com/pytorch/pytorch/issues/159169, I am not sure how feasible it is to make it a test in CI though, as in CI the CPU core varies between runs and could interfering with other test parallelizations.

zpcore replied:
Hah, I was also looking into this example. I didn't realize that the CPU core can vary in the CI. Maybe we can upload some benchmarks and run offline to see the performance number for future PRs. I can take a further look.

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

zpcore commented:
> @zpcore sorry the example feels very confused to me, for `aten.sort` I think the `dim` should affect the sharding strategy output, why it would not affect the sharding output?

This is just an example using `aten.sort`. Assume I am lazy to come up with a fancy sharding strategy, and just apply the replication strategy (all input/output replicated) for `aten.sort`, then the input `dim` should not impact this strategy output. So we use `register_op_strategy(replicate_op_strategy)(aten.sort.default)` without schema_info to register the strategy. 

However, this can produce the incorrect output because the original code uses the incorrect `dim` value from previous hashing. That can be the reason why this PR can fix some of the coverage test ops because those ops takes incorrect args from previous run due to strategy hashing.

XilunWu commented on torch/distributed/tensor/_op_schema.py:
```diff
@@ -562,6 +562,7 @@ class OutputSharding:
     output_spec: OutputSpecType
     redistribute_schema: Optional[OpSchema] = None
     needs_redistribute: bool = False
+    use_val_from_redistribute_schema: bool = False
```
add comment to OutputSharding about what "use_val_from_redistribute_schema" is for.

zpcore replied:
Done!

XilunWu reviewed (approved):
LGTM, thanks for fixing!

zpcore committed:
Update on "fix strategy hashing arg mismatch"

Strategy hashing output regarding an operator should only depend on the operator's input variable that can potentially change the output sharding strategy. (correct me if I am wrong...)
However, what I notice is that even though a variable doesn't impact the strategy, the hashing can still be hit and causing the variable to be reused. This will produce incorrect result, unless we specify `schema_info=RuntimeSchemaInfo(static_argnum=..., static_kwargkey=...)`.

Taking the example `test_call_with_different_nontensor_args` added in this PR, if we change the op_info when we call torch.sort the second time, we can see:

```
(Pdb) up
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
(Pdb) l
150             # extract local tensor and sharding infos to a OpInfo
151             op_info = self.unwrap_to_op_info(op_call, args, kwargs)
152             logger.debug("Dispatching op_call: %s", op_info.schema)
153  
154             self.sharding_propagator.propagate(op_info)
155  ->         torch.distributed.breakpoint()
156             output_sharding = op_info.output_sharding
157             logger.debug("output_sharding for %s: %s", op_call, output_sharding)
158             assert output_sharding is not None, "output sharding should not be None"
159  
160             mesh = op_info.compute_mesh

(Pdb) w
...
  /data/users/piz/pytorch/test/distributed/tensor/test_op_strategy.py(627)test_call_with_different_nontensor_args()
-> out1, _ = torch.sort(sharded_dtensor, dim=1)  # sort each row
  /data/users/piz/pytorch/torch/_compile.py(53)inner()
-> return disable_fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/_dynamo/eval_frame.py(1005)_fn()
-> return fn(*args, **kwargs)
  /data/users/piz/pytorch/torch/distributed/tensor/_api.py(358)__torch_dispatch__()
-> return DTensor._op_dispatcher.dispatch(
> /data/users/piz/pytorch/torch/distributed/tensor/_dispatch.py(155)dispatch()
-> torch.distributed.breakpoint()
  /data/users/piz/pytorch/torch/distributed/__init__.py(119)breakpoint()
-> meta_in_tls = torch._C._meta_in_tls_dispatch_include()

(Pdb) op_info
OpInfo(
    compute_mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), 
    schema=OpSchema(op=aten.sort.default, args_schema=(Spec(S(1) on (4, 4)), 1), kwargs_schema={}), 
    flat_args_schema=[DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Shard(dim=1),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 1], 
    local_args=(tensor([[29.],
        [25.],
        [ 1.],
        [48.]], device='cuda:0'), 1), 
    local_kwargs={}, 
    args_tree_spec=None, 
    output_sharding=OutputSharding(
        output_spec=(
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.float32)), 
            DTensorSpec(mesh=DeviceMesh((4,), device: 'cuda', stride: (1,)), placements=(Replicate(),), tensor_meta=TensorMeta(shape=torch.Size([4, 4]), stride=(4, 1), dtype=torch.int64))), 
            redistribute_schema=OpSchema(
                op=aten.sort.default, 
                args_schema=(Spec(R on (4, 4)), 0), 
                kwargs_schema={}), 
            needs_redistribute=True)
)
```
In the op_info, we can see that even though `dim=1` is the input, output_sharding.output_spec.redistribute_schema.args_schema still uses `dim=0` since this output_sharding is cached from the first torch.sort call. However, the true dim value is still preserved in the op_info.schema.args_schema. I believe this is the one we should use.


I think this fix will 1) improve the chance of get a strategy cache hit. 2) Remove the requirement of static_args in RuntimeSchemaInfo(1) for some ops. E.g, for op `histc`, the test should pass now.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]
